packages/grub/0045-mkimage-pgp-move-single-public-key-into-its-own-sect.patch

```diff
@@ -0,0 +1,261 @@
+From eaa371d26be8f76f9d7ce7e9c4b5d30684a48f11 Mon Sep 17 00:00:00 2001
+From: Markus Boehme <markubo@amazon.com>
+Date: Mon, 24 Oct 2022 15:20:23 +0000
+Subject: [PATCH] mkimage, pgp: move single public key into its own section of
+ EFI image
+
+If mkimage is asked to embed a single public key for signature checking
+into an EFI image, move that key into a new .pubkey section of the PE
+file. Moving the key into its dedicated section allows for easily
+swapping the public key, e.g. using objcopy, without having to rebuild
+the image.
+
+If more than one key is to be embedded, no new section is created and
+all keys are embedded as modules just as before. The PGP signature
+verification will check both of these sources for valid keys.
+
+Signed-off-by: Markus Boehme <markubo@amazon.com>
+---
+ grub-core/commands/pgp.c | 28 ++++++++++++++
+ include/grub/efi/efi.h   |  2 +-
+ util/mkimage.c           | 84 +++++++++++++++++++++++++++-------------
+ 3 files changed, 87 insertions(+), 27 deletions(-)
+
+diff --git a/grub-core/commands/pgp.c b/grub-core/commands/pgp.c
+index b81ac0a..5fa1e8e 100644
+--- a/grub-core/commands/pgp.c
++++ b/grub-core/commands/pgp.c
+@@ -32,6 +32,7 @@
+ #include <grub/kernel.h>
+ #include <grub/extcmd.h>
+ #include <grub/verify.h>
++#include <grub/efi/efi.h>
+ 
+ GRUB_MOD_LICENSE ("GPLv3+");
+ 
+@@ -959,6 +960,33 @@ GRUB_MOD_INIT(pgp)
+     grub_pk_trusted = pk;
+   }
+ 
++#ifdef GRUB_MACHINE_EFI
++  {
++    grub_addr_t pubkey_section;
++    grub_uint32_t pubkey_vsz;
++
++    pubkey_section = grub_efi_section_addr (".pubkey", &pubkey_vsz);
++    if (pubkey_section)
++      {
++	struct grub_file pseudo_file;
++	struct grub_public_key *pk;
++
++	grub_memset (&pseudo_file, 0, sizeof (pseudo_file));
++
++	pseudo_file.fs = &pseudo_fs;
++	pseudo_file.size = pubkey_vsz;
++	pseudo_file.data = (char *) pubkey_section;
++
++	pk = grub_load_public_key (&pseudo_file);
++	if (!pk)
++	  grub_fatal ("error loading key from .pubkey section: %s\n", grub_errmsg);
++
++	pk->next = grub_pk_trusted;
++	grub_pk_trusted = pk;
++      }
```

Comment:
Hmm, I think these tabs in `pgp.c` might be correct (per GNU style) since the `if` statement increases the indent to 8 spaces, but `mkimage.c` doesn't seem to match the style.

Reply:
Unless you feel very strongly about this, I'm inclined to just leave the patch to `mkimage.c` as-is. I tried a number of approaches - `indent` runs, vim retabs - and wasn't happy with the results.

Reply:
I do not feel strongly about it. Let's leave it as is. Checking random files in grub they all had mixed-indent so it does not really matter.

Reply:
Indeed, it's GNU style indentation. There are [mixed opinions](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/coding-style.rst?h=v6.4#n12) on it...